firedrake/slate/slate.py

```diff
@@ -0,0 +1,616 @@
+        elif self.rank == 2:
+            return "M_%d" % self.id
+
+    def __repr__(self):
+        """SLATE representation of the tensor object."""
+        if self.rank == 0:
+            return "Scalar(%r)" % self.form
+        elif self.rank == 1:
+            return "Vector(%r)" % self.form
+        elif self.rank == 2:
+            return "Matrix(%r)" % self.form
+
+
+class UnaryOp(TensorBase):
+    """An abstract SLATE class for representing unary operations on a
+    Tensor object.
+
+    The currently supported unary operations on a SLATE :class:`TensorBase`
+    expression are the following:
+
+        (1) the inverse of a tensor, `A.inv`, implemented in the subclass
+            `Inverse`;
+        (2) the tranpose of a tensor, `A.T`, implemented in the subclass
+            `Transpose`;
+        (3) and the negative operation, `-A` (subclass `Negative`).
+
+    :arg A: a :class:`TensorBase` object. This can be a terminal tensor object
+            (:class:`Tensor`) or any derived expression resulting from any number
+            of linear algebra operations on `Tensor` objects. For example,
+            another instance of a `UnaryOp` object is an acceptable input, or
+            a `BinaryOp` object.
+    """
+
+    def __init__(self, A):
+        """Constructor for the UnaryOp class."""
+        if not isinstance(A, TensorBase):
+            raise Exception("Expecting a `slate.TensorBase` object, not %r" % A)
+        self.tensor = A
+        self.check_operand_inversion(A)
+        super(UnaryOp, self).__init__()
+
+    def arguments(self):
+        """Returns a tuple of arguments associated with the tensor."""
+        return self.get_uop_arguments(self.tensor)
+
+    def coefficients(self):
+        """Returns a tuple of coefficients associated with the tensor."""
+        return self.tensor.coefficients()
+
+    def ufl_domains(self):
+        """Returns the integration domains of the integrals associated with
+        the tensor.
+        """
+        return self.tensor.ufl_domains()
+
+    def subdomain_data(self):
+        """Returns a mapping on the tensor:
+        `{domain:{integral_type: subdomain_data}}`.
+        """
+        return self.tensor.subdomain_data()
+
+    @classmethod
+    def check_operand_inversion(cls, A):
+        pass
+
+    @classmethod
+    def get_uop_arguments(cls, A):
+        """Returns the expected arguments of the resulting tensor of
+        performing a specific unary operation on a tensor.
+
+        Implemented in subclass.
+        """
+        pass
+
+    @property
+    def operands(self):
+        """Returns an iterable of the operands of this operation."""
+        return (self.tensor,)
+
+    def __str__(self, prec=None):
+        """String representation of a resulting tensor after a unary
+        operation is performed.
+        """
+        if self.op == "Inverse":
+            return "(%s).inv" % self.tensor
+
+        elif self.op == "Transpose":
```

Comment:
Note that if...elif...else is an OO anti-pattern. I suggested it for `Tensor.__str__` because that was the only different thing in `Scalar`, `Vector`, and `Tensor`, so we could merge those classes. But here you still have `Inverse`, `Transpose`, and `Negative` subclasses, it would be clearer to just define `__str__` in each.